test/deepClone.test.ts

```diff
@@ -0,0 +1,170 @@
+
+    // @ts-ignore
+    const object = { key: null };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `undefined` value", () => {
+
+    // @ts-ignore
+    const object = { key: undefined };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `boolean` value", () => {
+
+    // @ts-ignore
+    const object = { key: Boolean };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `string` value", () => {
+
+    // @ts-ignore
+    const object = { key: "hello world" };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `symbol` value", () => {
+
+    // @ts-ignore
+    const object = { key: Symbol("a") };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `func` value", () => {
+
+    // @ts-ignore
+    const object = { key() { console.log("hello"); } };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `array` value", () => {
+
+    // @ts-ignore
+    const object = { key: [[1], [2], [3], [4]] };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `object1` value", () => {
+
+    // @ts-ignore
+    const object = { key: { a: 1, b: 2, c: 3 } };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `object2` value", () => {
+
+    // @ts-ignore
+    const object = { key: { a: { a1: 1 }, b: { b1: 1 }, c: { a1: 1 }, d: ["hello", "world", ["!!"]] } };
+
+    const expected = object;
+    const actual = deepClone(object);
+    const test = isCloned(expected, actual);
+
+    expect(expected).toEqual(actual);
+    expect(test).toBe(true);
+});
+
+test("should clone object with `date` value", () => {
+
+    // @ts-ignore
+    const object: any = { key: new Date() };
```

Comment:
In my tests (not unit tests, just NodeJS REPL from the command line), `Date` object becomes `{}` when cloned with the `deepClone()` function. Obviously, this doesn't matter when inputing data from raw JSON (where `Date` is not a primitive type), but the clone function should nonetheless not loose the stringified date?

Reply:
I used both `JSON.stringify()` and `util.inspect()` to discover this Date bug. Here is my test script:
```
const util = require("util");

const deepClone = (o) =>
    o && typeof o === "object"
        ? Array.isArray(o)
            ? o.map((value) => deepClone(value))
            : Object.fromEntries(
                Object.entries(o).map(
                    ([key, value]) =>
                        [key, deepClone(value)]))
        : o;
const deepClone_ = (o) =>
    o && typeof o === "object"
        ? Array.isArray(o)
            ? o.map((value) => deepClone_(value))
            : Object.entries(o).reduce((obj, [key, value]) => ({...obj, [key]: deepClone(value)}), {})
        : o;
const obj = {
    zund: undefined,
    zfunc1: (...args) => { console.log("hello", args); },
    zfunc2(...args) { console.log("hello", args); },
    zdate: new Date(),
    zbool: true,
    zemptyarr: [],
    zstring: "str1",
    znul: null,
    zsym: Symbol("zzz"),
    znumber: -1,
    zarray: [true, "str3", 111, {foo: "bar", deep: ["another array"]}, [[], 1, 2, 3, {n:4}]],
    zobj: {
        zbool: false,
        zstring: "str2",
        znumber: 999,
        zarray: [false, "str5", 333, {num: 555, zsymb: Symbol("yyy"), nest: [{nul: null, und: undefined}]}, [1, 2, 3, {n:4}]],
    },
};
console.log(JSON.stringify(obj, null, 4));
console.log(util.inspect(obj, { colors: true, depth: null, compact: false }));

console.log(JSON.stringify(deepClone(obj), null, 4));
console.log(util.inspect(deepClone(obj), { colors: true, depth: null, compact: false }));

console.log(JSON.stringify(deepClone_(obj), null, 4));
console.log(util.inspect(deepClone_(obj), { colors: true, depth: null, compact: false }));

```

Reply:
i added a shallow copy by their copy constructor of Date, Set, Map